### ISSUE TITLE
Caller is requesting unknown variable: hc_concentration

### ISSUE THREAD

**Issue by bpbond** [JGCRI/hector#372]
Something goofy has developed in the Hector command-line build; maybe it's a code problem, maybe it's an instructions problem, but either way it's a problem. I'm concerned that with all the focus on the R package over the last 1-2 years we've neglected the core (command line) configuration.

> Well, Dominique, I owe you an apology. This morning I built the Hector master, updating the Xcode project settings (see PR Kalyn), and initially got the same error—
>  
> Caller is requesting unknown variable: hc_concentration
>  
> I didn’t understand this; the R and command-line versions are built from the same sources, and if the former runs, the latter should as well.
>  
> BUT after quitting and then restarting Xcode, and rebuilding everything…Hector ran without a problem:
>  
> Thu May 21 05:33:17 2020:NOTICE:run_spinup: Carbon model is spun up after 352 steps
> Thu May 21 05:33:17 2020:NOTICE:run_spinup: Model spun up after 352 steps
> Thu May 21 05:33:17 2020:NOTICE:main: Running the core.
> Thu May 21 05:33:17 2020:NOTICE:run: Running...
> Thu May 21 05:33:19 2020:NOTICE:main: Hector wrapper end
> Program ended with exit code: 0
>  
> Good, but now I really don’t understand! We definitely have a problem, even though I’m not yet sure what it is.
>  
> Compounding things, I took a careful look at the Travis logs, and test_hector.sh is broken but isn’t generating the error code that would have caused Travis to fail and thus alert us. [ this is #371 ]

We need to figure out what's going on.

**Comment by rgieseke** [comment]
This one seems to introduce the error: af1c8023290f2e7d8496f74b60d7297ed11e4eb4

Or the previous WIP one (which doesn't compile): 6e0527dd20b32acfafc1dceaf34ca65f9897d813

**Comment by rgieseke** [comment]
Adding back this fixes this: https://github.com/JGCRI/hector/commit/af1c8023290f2e7d8496f74b60d7297ed11e4eb4#diff-44737671487fd4ba4d4f8b817b217f20L213

```cpp
else if( varName == D_HC_CONCENTRATION ) {
        returnval = Ha_ts.get(getdate);
    }
```
But then i get a problem with the ocean component:
```
* Program exception:
msg:  	Assertion failed: Date required for ocean C flux
func: 	getData
file: 	ocean_component.cpp
ffile:	ocean_component.cpp

line: 	343
```

**Comment by bpbond** [comment]
Thanks @rgieseke . I'm going to make some time later this week to get back to this and dive in.

**Comment by bpbond** [comment]
Tagging @mensbrugghe so he can follow

**Comment by bpbond** [comment]
Yes this is happening in very first year of spinup, when the `csvOutputStreamVisitor` attempts to get the value for "CF4_halocarbon".

As @rgieseke correctly points out above, the code to handle this data request was accidentally deleted in 6e0527d.

**Why didn't the automated testing immediately see this error?**
1. For command-line Hector, there was a mistake in the command-line test script and it wasn't returning an error code as it should have. I have addressed this in #373 .
2. For R package Hector, the package tests only run with `new_core()` defaults, i.e. `suppress logging = TRUE`, _suppressing all logging and outputs_. So this error was never triggered, because the `csvOutputStreamVisitor` was never called. I have opened #381 for this.

**Comment by bpbond** [comment]
After fixing this there's a second problem being exposed: 

```
Thu Jun 11 10:53:37 2020:NOTICE:prepareToRun: Spinning up model...
* Program exception:
msg:  	Assertion failed: Date required for ocean C flux
func: 	getData
file: 	ocean_component.cpp
ffile:	/Users/d3x290/Dropbox/Documents/Work/Code/hector/src/ocean_component.cpp

line: 	343
```

On first inspection this relates to ab5537c by @ashiklom . It looks like the same thing happened: a commit that worked for the R version, but not the command line version. More in a bit.

And a third. I haven't checked but presumably from same commit(s).
```
* Program exception:
msg:  	Assertion failed: Date required for atmosphere -> land C flux
func: 	getData
file: 	simpleNbox.cpp
ffile:	/Users/d3x290/Dropbox/Documents/Work/Code/hector/src/simpleNbox.cpp

line: 	521
```

For each of these cases, I'm changing the logic so that if no date is given (i.e., we're in spinup) the current value is returned; if a date _is_  given, we pull it out of the time series.
